### PR TITLE
Issue248 doctest

### LINKED ISSUE #248: Clean up how to handle Catch
Two related problems:

1. Catch2 is probably the least common library that we use. Since it is—as we use it—just a single header, it should be easy for a user building the model to specify the location of the Catch 2 header and for it to be located by CMake.
2. The modern version of Catch2 is not the single header version any more.

## Adding a non-system header
One way of specifying the location is to follow the commands on [this page](https://cliutils.gitlab.io/modern-cmake/chapters/testing/catch.html) under the heading 'Vendoring' (whatever that means).

```
set(CATCH_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/extern/include)
add_library(Catch2::Catch2 IMPORTED INTERFACE)
set_property(TARGET Catch2::Catch2 PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${CATCH_INCLUDE_DIR}")
```
where the header itself is at `extern/include/catch2/catch.hpp`.

## Replacing Catch
The header-only library [doctest](https://github.com/doctest/doctest) seems to be a replacement for the single header version of Catch2. It is available on Homebrew, at least, and is available in Ubuntu Jammy Jellyfish as `doctest-dev`. Even so, I would guess it is even less commonly available on HPC than Catch2, so we would need a robust solution to ensure it was always available. As doctest is MIT-licensed, this might be to include it in the source distribution by default.

### PR DESCRIPTION
**Pull request by timspainNERSC**
Replace Catch2 with doctest.

Closes #248

### CODE REVIEW

**Review by a-smith-github — CHANGES_REQUESTED** [review]
Thanks Tim,

This looks a nice swap. Doctest looks like a good call.

Could you expand in the PR about the change you made & it's limitations so we can refer to it later? I've made some assumptions in my review / had to investigate. Also some discussion about including the doctest header here or elsewhere for later reference. Also I don't think the issue captured why the decision to ship a copy of doctest was made, so that should be included here. 

A couple of suggestions/comments (new scheme for test naming and adding description, inclusion of Catch2 in CMake, follow-up issue tracking). 

One more thing. The test output looks like:

===============================================================================
[doctest] test cases:  2 |  2 passed | 0 failed | 0 skipped
[doctest] assertions: 36 | 36 passed | 0 failed |
[doctest] Status: SUCCESS!
[doctest] doctest version is "2.4.10"
[doctest] run with "--help" for options

But could we investigate modifying this so it declares the test name that ran (and the suit name if we go down that route)? --success=true looks overkill maybe leveraging --duration=true (this could also help us track performance regressions going forward).  Also what does a failure look like? --- this should be captured in the PR description / design doc. Obviously this helps diagnosing issues but potentially allows us to track failures better, monitor sporadic issues (which we might start to face more of as we move to multi core and multi process computations and IO), performance issues. 

In general though really happy with doctest & nice work

File: CMakeLists.txt [hunk 1]
```diff
@@ -43,6 +43,11 @@ add_definitions(-DBOOST_ALL_DYN_LINK)
 find_package(Boost COMPONENTS program_options log REQUIRED)
 
 find_package(Catch2 REQUIRED)
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply CMakeLists.txt:45]: No, I just missed doing the dynamics/test directory along with the other tests.

[the remark this answers is not in the document]
Comment (timspainNERSC) [reply CMakeLists.txt]: Replaced with doctest.

**Comment by timspainNERSC** [comment]
What does a doctest failure look like?
```
TEST CASE:  ConfiguredAtmosphere melting test

/Users/timpai/nextsim/dg/nextsimdg/physics/test/ConfiguredAtmosphere_test.cpp:137: FATAL ERROR: REQUIRE( qia[0] == doctest::Approx(-83.5952).epsilon(prec) ) is NOT correct!
  values: REQUIRE( -84.5952 == Approx( -83.5952 ) )

===============================================================================
[doctest] test cases: 1 | 0 passed | 1 failed | 0 skipped
[doctest] assertions: 2 | 1 passed | 1 failed |
[doctest] Status: FAILURE!
```

File: core/test/Iterator_test.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ size_t zeroTime<size_t>()
 }
 
 
-TEST_CASE("Count iterator testing", "[Iterator]")
+TEST_CASE("Count iterator testing")
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply core/test/Iterator_test.cpp:50]: I have wrapped all the TEST_CASEs in a TEST_SUITE (per file) that takes the name of the class under test. To me that reproduces the original functionality, and is sufficient for this PR. Any tests that needs a more detailed description can be added to an issue. (**I** think they're fine.)

File: lib/doctest/doctest.h
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply lib/doctest/doctest.h:1]: I have created issue #255 to create a release checklist. Currently the only item is to check and update doctest as necessary.

**Comment by timspainNERSC** [comment]
You are of course quite right about the doctest LICENSE.txt, which is now included.

**Review by a-smith-github — APPROVED** [review]

**Pull request merged**